arangod/Agency/Supervision.cpp

```diff
@@ -1079,6 +1108,163 @@ void Supervision::handleShutdown() {
   }
 }
 
+
+struct ServerLists {
+  std::unordered_map<ServerID, uint64_t> dbservers;
+  std::unordered_map<ServerID, uint64_t> coordinators;
+};
+
+//  Get all planned servers
+//  If heartbeat in transient too old or missing
+//    If heartbeat in snapshot older than 1 day
+//      Remove coordinator everywhere
+//      Remove DB server everywhere, if not leader of a shard
+
+std::unordered_map<ServerID, std::string> deletionCandidates (
+  Node const& snapshot, Node const& transient, std::string const& type) {
+  using namespace std::chrono;
+  std::unordered_map<ServerID, std::string> serverList;
+  std::string const planPath = "/Plan/" + type;
+
+  if (snapshot.has(planPath) && !snapshot(planPath).children().empty()) {
+
+    std::string persistedTimeStamp;
+
+    for (auto const& serverId : snapshot(planPath).children()) {
+      auto const& transientHeartbeat =
+        transient.hasAsNode("/Supervision/Health/" + serverId.first);
+      try {
+        // Do we have a transient heartbeat younger than a day?
+        if (transientHeartbeat.second) {
+          auto const t = stringToTimepoint(
+            transientHeartbeat.first("Timestamp").getString());
+          if (t > system_clock::now() - hours(24)) {
+            continue;
+          }
+        }
+        // Else do we have a persistent heartbeat younger than a day?
+        auto const& persistentHeartbeat =
+          snapshot.hasAsNode("/Supervision/Health/" + serverId.first);
+        if (persistentHeartbeat.second) {
+          persistedTimeStamp = persistentHeartbeat.first("Timestamp").getString();
+          auto const t = stringToTimepoint(persistedTimeStamp);
+          if (t > system_clock::now() - hours(24)) {
+            continue;
+          }
+        } else {
+          if(!persistedTimeStamp.empty()) {
+            persistedTimeStamp.clear();
+          }
+        }
+      } catch (std::exception const& e) {
+        LOG_TOPIC("21a9e", DEBUG, Logger::SUPERVISION)
+          << "Failing to analyse " << serverId << " as deletion candidate " << e.what();
+      }
+
+      // We are still here?
+      serverList.emplace(std::pair<ServerID, std::string>{serverId.first, persistedTimeStamp});
```

Comment:
This is suboptimal: A pair is constructed and then copy-constructed in the container. I suggest to leave out the explicit `std::pair` construction.

Reply:
How about `try_emplace` or `std::piecewise_construct`